src/compiler/compiler.rs

```diff
@@ -332,32 +334,45 @@ fn dist_or_local_compile<T>(dist_client: Arc<dist::Client>,
                             cwd: PathBuf,
                             compilation: Box<Compilation>,
                             weak_toolchain_key: String,
-                            toolchain_creator: Box<FnMut(File) -> io::Result<()>>,
+                            toolchain_creator: Box<CompilerPackager>,
                             out_pretty: String)
                             -> SFuture<(Cacheable, process::Output)>
         where T: CommandCreatorSync {
+    use boxfnonce::BoxFnOnce;
     use futures::future;
 
     debug!("[{}]: Attempting distributed compilation", out_pretty);
     let compile_out_pretty = out_pretty.clone();
     let compile_out_pretty2 = out_pretty.clone();
-    let (compile_cmd, dist_compile_cmd, cacheable) = compilation.generate_compile_commands().unwrap();
+    let compile_out_pretty3 = out_pretty.clone();
+    let mut path_transformer = dist::PathTransformer::new();
+    let (compile_cmd, dist_compile_cmd, cacheable) = compilation.generate_compile_commands(&mut path_transformer).unwrap();
+    let local_executable = compile_cmd.executable.clone();
     // TODO: the number of map_errs is subideal, but there's no futures-based carrier trait AFAIK
     Box::new(future::result(dist_compile_cmd.ok_or_else(|| "Could not create distributed compile command".into()))
         .and_then(move |dist_compile_cmd| {
-            let output_paths = compilation.outputs().map(|(_key, path)| cwd.join(path)).collect();
-            // TODO: does this need to be in future::result?
-            compilation.into_inputs_creator()
-                .map(|inputs_creator| (dist_compile_cmd, inputs_creator, output_paths))
+            debug!("[{}]: Creating distributed compile request", compile_out_pretty);
+            let dist_output_paths = compilation.outputs()
+                .map(|(_key, path)| path_transformer.to_dist_abs(&cwd.join(path)))
+                .collect::<Option<_>>()
+                .unwrap();
+            compilation.into_dist_inputs_creator(&mut path_transformer)
+                .map(|dist_inputs_creator| (path_transformer, dist_compile_cmd, dist_inputs_creator, dist_output_paths))
         })
-        .and_then(move |(dist_compile_cmd, inputs_creator, output_paths)| {
-            debug!("[{}]: Distributed compile request created, requesting allocation", compile_out_pretty);
+        .and_then(move |(path_transformer, mut dist_compile_cmd, dist_inputs_creator, dist_output_paths)| {
+            debug!("[{}]: Identifying toolchain", compile_out_pretty2);
+            let toolchain_creator_cb = BoxFnOnce::from(move |f| toolchain_creator.write_pkg(f));
             // TODO: put on a thread
```

Comment:
If you don't have a fix for this TODO in a later patch, can you file a follow-up issue for it so we don't forget?